### PR TITLE
feat(#108): support custom register provider

### LINKED ISSUE #108: [Question] How to use custom RegisterProvider ?
This plugin will add a `@RegisterProvider(CompositeAuthenticationProvider)` annotation at generated Api interface,   when authentication is defined in openapi spec.  

But in some case, we need custom RegisterProvider, e.g. logging. I wonder how to achieve it.

### PR DESCRIPTION
**Pull request by Leibnizhu**
close #108

### CODE REVIEW

File: integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Leibniz on 2022/7/19.
+ */
+public class CountProvider implements ClientRequestFilter {
```
Comment (ricardozanini) [inline integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java:17]: Can you please remove this comment?

**Comment by ricardozanini** [comment]
I'll do an in-depth review later.

File: README.md [hunk 2]
```diff
@@ -537,6 +537,44 @@ file. By default, these attributes are generated. You can fine tune this behavio
...
 
+## Custom Register Providers for generated api
+
+In some cases, we need custom `RegisterProvider` for generated api, e.g. logging. You can define your own Providers in `application.properties` :
+
+| Property Key                                                                                                                                                                        | Example                                                                                                                                |
+|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
+| `quarkus.openapi-generator.codegen.spec.[filename].custom-register-providers` | `quarkus.openapi-generator.codegen.spec.simple_openapi_json.custom-register-providers=org.test.Foo,org.test.Bar`<br/>Provider classes are separated by commas |
+
+With the above configuration, your Rest Clients will be created with a code similar to the following:
```
Comment (ricardozanini) [inline README.md:548]:
```suggestion
With the above configuration, the extension generates your Rest Clients with a code similar to the following:
```

**Review by ricardozanini — APPROVED** [review]
Great work! Many thanks! There's one minor nit. :)

**Review by fjtirado — APPROVED** [review]

File: integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CountProvider implements ClientRequestFilter {
+
+    private final Logger log = LoggerFactory.getLogger(CountProvider.class);
```
Comment (fjtirado) [inline integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java:16]: This should be static final

File: integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Override
+    public void filter(ClientRequestContext context) throws IOException {
+        String operationName = String.join(":", context.getMethod(), context.getUri().toString());
+        REQUEST_COUNT.computeIfPresent(operationName, (op, count) -> count + 1);
+        log.info("Request {} for {} times", operationName, REQUEST_COUNT.get(operationName));
+    }
+
+    public Map<String, Integer> requestCounts() {
+        return new HashMap<>(REQUEST_COUNT);
```
Comment (fjtirado) [inline integration-tests/example-project/src/main/java/org/acme/openapi/weather/api/provider/CountProvider.java:27]: Collections.unmodifiableMap() ratherr than a new map

**Review by fjtirado — COMMENTED** [review]
Just minor things

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
@Leibnizhu let me know if you need a new release for this feature.

File: deployment/src/main/resources/templates/api.qute [hunk 5]
```diff
@@ -39,6 +39,9 @@ import io.quarkiverse.openapi.generator.annotations.GeneratedParam;
 @RegisterProvider(CompositeAuthenticationProvider.class)
 @RegisterClientHeaders(AuthenticationPropagationHeadersFactory.class)
 {/if}
+{#for crpClassConfig in custom-register-providers.orEmpty}
```
Comment (YassinHajaj) [inline deployment/src/main/resources/templates/api.qute:42]: Hello,
Isn't the import missing here ?

Comment (ricardozanini) [reply deployment/src/main/resources/templates/api.qute:42]: Hi @YassinHajaj! Thanks for pointing this out. This information came from the property the user configured, so it must be the FQN. There's nothing we can do about it here. See: https://github.com/quarkiverse/quarkus-openapi-generator#custom-register-providers-for-generated-api

**Pull request merged**